File: medium/four_sum_count.rb
#https://leetcode.com/problems/4sum-ii/
#454. 4Sum II

def four_sum_count(a, b, c, d)
    result = 0
    size = a.size-1
    first = Hash.new(0)
    (0..size).each do |xa|
        (0..size).each do |xb|
           first[a[xa] + b[xb]] += 1
        end
    end
    (0..size).each do |xc|
        (0..size).each do |xd|
            result += first[-(c[xc]+d[xd])] if first.key?(-(c[xc]+d[xd]))
        end
    end
    result
end

File: medium/FourSumCount.java
//https://leetcode.com/problems/4sum-ii/
//454. 4Sum II

class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int result = 0;
        Map<Integer, Integer> ab_map = new HashMap<Integer, Integer>();
        for(int a:A){
            for(int b:B){
                ab_map.put(a+b,ab_map.getOrDefault(a+b,0) + 1);
            }
        }

        for(int c:C){
            for(int d:D){
                result += ab_map.getOrDefault(-(c+d), 0);
            }
        }
        return result;
    }
}